File: app/src/main/java/com/example/trigger/QRShowActivity.java
package com.example.trigger;

import android.graphics.Bitmap;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.ImageView;
import android.widget.Toast;

import com.google.zxing.BarcodeFormat;
import com.google.zxing.MultiFormatWriter;
import com.google.zxing.WriterException;
import com.google.zxing.common.BitMatrix;
import com.journeyapps.barcodescanner.BarcodeEncoder;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;


public class QRShowActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_qrshow);

        int setup_id = getIntent().getIntExtra("setup_id", -1);
        Setup setup = Settings.getSetup(setup_id);

        if (setup != null) {
           setTitle(getTitle() + ": " + setup.getName());

            try {
                generateQR(setup);
            } catch (Exception e) {
                e.printStackTrace();
                Toast.makeText(this, e.getMessage(), Toast.LENGTH_LONG).show();
            }
        } else {
            Toast.makeText(this, "Setup not found.", Toast.LENGTH_LONG).show();
        }
    }

    private Collection<String> collect(Iterator<String> iter) {
        ArrayList<String> set = new ArrayList<String>();
        while (iter.hasNext()) {
            set.add(iter.next());
        }
        return set;
    }

    ArrayList<String> getJsonKeys(JSONObject obj) {
        ArrayList<String> keys = new ArrayList<String>();
        Iterator<String> it = obj.keys();
        while (it.hasNext()) {
            keys.add(it.next());
        }
        return keys;
    }

    private String encodeSetup(JSONObject obj) {
        // do not export internal id
        obj.remove("id");

        // remove empty strings, images and null values
        ArrayList<String> keys = getJsonKeys(obj);
        for (String key : keys) {
            Object value = obj.opt(key);
            if (value == null) {
                obj.remove(key);
            } else if (key.endsWith("_image")) {
                obj.remove(key);
            } else if (value instanceof String) {
                String s = (String) value;
                if (s.length() == 0) {
                    obj.remove(key);
                }
            }
        }

        return obj.toString();
    }

    private void generateQR(Setup setup) throws Exception {
        MultiFormatWriter multiFormatWriter = new MultiFormatWriter();
        int data_length = 0;
        try {
            JSONObject obj = Settings.toJsonObject(setup);
            String data = encodeSetup(obj);
            data_length = data.length();

            // data has to be a string
            BitMatrix bitMatrix = multiFormatWriter.encode(data, BarcodeFormat.QR_CODE, 1080, 1080);
            BarcodeEncoder barcodeEncoder = new BarcodeEncoder();
            Bitmap bitmap = barcodeEncoder.createBitmap(bitMatrix);
            ((ImageView) findViewById(R.id.QRView)).setImageBitmap(bitmap);
        } catch (WriterException e) {
            Toast.makeText(this, e.getMessage() + " (" + data_length + " Bytes)", Toast.LENGTH_LONG).show();
            finish();
        }
    }
}

File: app/src/main/kotlin/app/trigger/QRShowActivity.kt
/*
* Copyright (C) 2025 The Trigger Contributors
* SPDX-License-Identifier: GPL-3.0-or-later
*/

package app.trigger

import android.os.Bundle
import android.widget.ImageView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
import com.google.zxing.MultiFormatWriter
import com.google.zxing.WriterException
import com.journeyapps.barcodescanner.BarcodeEncoder
import org.json.JSONObject

class QRShowActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_qrshow)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        val door_id = intent.getIntExtra("door_id", -1)
        val door = Settings.getDoor(door_id)
        if (door != null) {
            title = "$title: ${door.name}"
            try {
                generateQR(door)
            } catch (e: Exception) {
                e.printStackTrace()
                Toast.makeText(this, e.message, Toast.LENGTH_LONG).show()
            }
        } else {
            Toast.makeText(this, "Setup not found.", Toast.LENGTH_LONG).show()
        }
    }

    private fun getJsonKeys(obj: JSONObject): ArrayList<String> {
        val keys = ArrayList<String>()
        val it = obj.keys()
        while (it.hasNext()) {
            keys.add(it.next())
        }
        return keys
    }

    private fun encodeSetup(obj: JSONObject): String {
        // do not export internal id
        obj.remove("id")

        // remove empty strings, images and null values
        val keys = getJsonKeys(obj)
        for (key in keys) {
            val value = obj.opt(key)
            if (value == null) {
                obj.remove(key)
            } else if (key.endsWith("_image")) {
                obj.remove(key)
            } else if (value is String) {
                if (value.length == 0) {
                    obj.remove(key)
                }
            }
        }
        return obj.toString()
    }

    private fun generateQR(door: Door) {
        val multiFormatWriter = MultiFormatWriter()
        var data_length = 0
        try {
            val obj = Settings.toJsonObject(door) ?: throw Exception("Failed to convert setup to JSON")
            val data = encodeSetup(obj)
            data_length = data.length

            // data has to be a string
            val hints =  mapOf(EncodeHintType.CHARACTER_SET to "utf-8")
            val bitMatrix = multiFormatWriter.encode(data, BarcodeFormat.QR_CODE, 1080, 1080, hints)
            val barcodeEncoder = BarcodeEncoder()
            val bitmap = barcodeEncoder.createBitmap(bitMatrix)
            findViewById<ImageView>(R.id.QRView).setImageBitmap(bitmap)
        } catch (e: WriterException) {
            Toast.makeText(this, "${e.message} ($data_length Bytes)", Toast.LENGTH_LONG).show()
            finish()
        }
    }

    companion object {
        private const val TAG = "QRShowActivity"
    }
}
